Computationally stiff tests

edrumwri opened the pull request:
Additional tests for the implicit integrator that really stress computational stiffness. Was used to uncover (and address) a flaw in the newly landed implicit Euler integrator.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6228)
<!-- Reviewable:end -->

edrumwri commented:
+@amcastro-tri for feature review please

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-:-Kl_EsjeHi3HSqlIc8HC:brbsdox)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Just a few comments to be addressed or add some docs on. With that, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 6 of 18 files at r7.
Review status: all files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[drake/examples/rod2d/BUILD, line 45 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kldhmd-wqBWc39Wn_Dz:-Kldhmd-wqBWc39Wn_E-:btwds9d) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/examples/rod2d/BUILD#L45)):*
> ```Python
>         "//drake/common",
>         "//drake/solvers:mathematical_program",
>         "//drake/systems/analysis",
> ```

why this new dependency? 

---

*[drake/systems/analysis/BUILD, line 107 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldZgJyDzn8BdazZ-0v:-KldZgJyDzn8BdazZ-0w:btcjt03) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/BUILD#L107)):*
> ```Python
>     hdrs = ["test/robertson_system.h"],
>     deps = [
>     ],
> ```

BTW, line break not needed.

---

*[drake/systems/analysis/BUILD, line 114 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldZiLJwz8pfGmlcRdK:-KldZiLJwz8pfGmlcRdL:btcjt03) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/BUILD#L114)):*
> ```Python
>     testonly = 1,
>     hdrs = ["test/stiff_double_mass_spring_system.h"],
>     deps = [
> ```

BTW, line break not needed.

---

*[drake/systems/analysis/implicit_euler_integrator-inl.h, line 381 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldWKpqaei5ZTxmuUhr:-KldWKpqaei5ZTxmuUhs:b2je09h) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/implicit_euler_integrator-inl.h#L381)):*
> ```objc
>     // this check, the convergence criteria can be applied when
>     // |dx_norm| ~ 1e-22 (one example taken from practice), which does not
>     // allow the norm to be reduced further.
> ```

BTW, this seems reasonable but I don't quite the comment you make here. Is this because theta below gets closer to one? (or even greater than one) due to fact that this norm becomes so small?
I wonder if this is what we want in the general case. If the units of `dx` correspond to large numerical values, then you might never achieve such a small norm even though for the units of the problem the norm of `dx` could be quite small. Would we need an absolute tolerance or something of the like here?

---

*[drake/systems/analysis/implicit_euler_integrator-inl.h, line 405 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldZ4rapfH3ijtYuKXn:-KldZ4rapfH3ijtYuKXo:b-97va8g) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/implicit_euler_integrator-inl.h#L405)):*
> ```objc
>       const double k_dot_tol = kappa * this->get_accuracy_in_use();
>       if (eta * dx_norm < k_dot_tol) {
>         SPDLOG_DEBUG(drake::log(), "Newton-Raphson converged; η = {}", eta);
> ```

BTW, I just realize that here you return successfully without updating the context's continuous state with the newer (hopefully better) `xtplus` (as you do later on in line 414. I think you'd like to have that update here also before you return.

---

*[drake/systems/analysis/implicit_euler_integrator-inl.h, line 470 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldYcvnV7YL83hyAtsE:-KldYcvo2UHx4ZzAw3s9:b-nw7jws) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/implicit_euler_integrator-inl.h#L470)):*
> ```objc
> //       integrator) are set to those at t0+dt on successful return (i.e., when
> //       the function returns `true`). State will be indeterminate on `false`
> //       return.
> ```

BTW, per your `@pre` and `@post` comments, I think a safer signature would take a mutable context (or mutable continuous state vector if you prefer) making this policies explicit in the method's signature.

---

*[drake/systems/analysis/implicit_euler_integrator-inl.h, line 473 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldYzkULqCC70h4ic6K:-KldYzkULqCC70h4ic6L:b-7lb7o6) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/implicit_euler_integrator-inl.h#L473)):*
> ```objc
> template <class T>
> bool ImplicitEulerIntegrator<T>::StepImplicitTrapezoid(const T& dt,
>                                                     const VectorX<T>& dx0,
> ```

BTW, odd indentation.

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 34 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldbtJDIYTmCktC8cq1:-KldbtJDIYTmCktC8cq2:b-s6mpro) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L34)):*
> ```cpp
>   state->SetAtIndex(0, 1);
>   state->SetAtIndex(1, 0);
>   state->SetAtIndex(2, 0);
> ```

Is the solution for t→∞ independent of the initial conditions?
I see the value of showing the integrator can go this far, but this test doesn't seem to test the accuracy of the solver (since any ok solver that is stable would be able to find the steady state solution).

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 37 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kldb5AJ9firzQL_Z1tQ:-Kldb5AJ9firzQL_Z1tR:bb4ru7i) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L37)):*
> ```cpp
> 
>   const double t_final = robertson->get_end_time();
>   const double tol = 5e-5;
> ```

BTW, notice that for this tolerance, the reference solution could've been taken to be `[0, 0 ,1]` and the test would still pass. This might indicate that a more stressful test would be one where the solution is non-trivial, like for instance at `t = 1`?

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 80 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldfXMgdv6YmvPC-blB:-KldfXMhbIQcSudT6772:btu6thn) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L80)):*
> ```cpp
> 
>     // Separate context necessary for the double spring mass system.
>     double_context_ = stiff_double_system_->CreateDefaultContext();
> ```

BTW, `double_context_` could for a someone like me be confused with the scalar type. Maybe rename to `double_spring_context_`?

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 254 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kldfo63Nvks5Uit6ZoF:-Kldfo64G7JemJEUQLri:bj7yaaf) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L254)):*
> ```cpp
> // and damper connecting two point masses together, and one of the point masses
> // is connected to "the world" using a spring with no damper. The solution of
> // this system should be identical to the solution of a undamped spring
> ```

BTW, **an** undamped

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 254 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldfyNFYLEtvznpxuKI:-KldfyNFYLEtvznpxuKJ:b75mwdo) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L254)):*
> The solution of this system should be identical ....

Are you thinking of a particular mode being excited here? or are did you mean to say that the solution **approximates** (not identical) that of the two masses moving together?

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 261 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldgEFmXGcRFY-BbEu2:-KldgEFmXGcRFY-BbEu3:bupina6) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L261)):*
> ```cpp
> 
>   // Designate the solution tolerance.
>   const double sol_tol = 2e-2;
> ```

could we use a smaller number? is this large because the target accuracy is large? or because the approximation to the two masses moving together is only good to this precision?

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 298 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kldhb-mKwRRGpD3-81d:-Kldhb-mKwRRGpD3-81e:b-hdyhn4) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L298)):*
> ```cpp
> TEST_F(ImplicitIntegratorTest, SpringMassDamperStiff) {
>   // Create a context.
>   auto context = spring_damper_->CreateDefaultContext();
> ```

not used?

---

*[drake/systems/analysis/test/robertson_system.h, line 9 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kld_ZmXeHa-sz0F5m6i:-Kld_ZmXeHa-sz0F5m6j:b-jnmqd6) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/robertson_system.h#L9)):*
> ```objc
> namespace drake {
> namespace systems {
> namespace analysis_test {
> ```

BTW, all throughout Drake we would've done:
```c++
namespace drake {
namespace systems {
namespace analysis {
namespace test {
```

I think worth trying to match wath everyone else does in Drake for consistency. 

---

*[drake/systems/analysis/test/robertson_system.h, line 13 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kld_3wpYW-i2GE_yLDo:-Kld_3wpYW-i2GE_yLDp:bjbk317) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/robertson_system.h#L13)):*
> ```objc
> /// Robertson's stiff chemical reaction problem. This example is taken from
> /// [Hairer, 1996] and is described in more detail in:
> /// http://www.radford.edu/~thompson/vodef90web/problems/demosnodislin/Single/DemoRobertson/demorobertson.pdf
> ```

BTW, I like this reference a lot  but I wonder if the site would be stable enough to have it here as a link. If the link dies then we lose this reference. Would there be a journal paper we can cite? could you push the pdf? Maybe worth complimenting stating the system of equations for this system and stating why is complex to solve (stiff, etc)
Ah! maybe we can cite the reference mentioned in that pdf? (H.H. Robertson in 1966 [Rob66])

---

*[drake/systems/analysis/test/stiff_double_mass_spring_system.h, line 14 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldcTIjeoa6RgbSM3Tk:-KldcTIjeoa6RgbSM3Tl:bdhtvz6) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/stiff_double_mass_spring_system.h#L14)):*
> ```objc
> /// This simple example is used to provide a stiff system for testing
> /// implicit integration. Mass cannot be set, nor can spring and damping
> /// constants be set.
> ```

BTW, a brief statement of the equations would be useful here, even in second order form if that makes them shorter.

---

*[drake/systems/analysis/test/stiff_double_mass_spring_system.h, line 37 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KlddXyAN155068HsJGi:-KlddXyAN155068HsJGj:byej5az) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/stiff_double_mass_spring_system.h#L37)):*
> ```objc
>     // Get the force from Spring 1 and Spring 2.
>     const T f1 = -k(0)*stretch0 - v(0)*b(0);
>     const T f2 = -k(1)*stretch1 - v(1)*b(1);
> ```

BTW, space around `*` operator.

---

*[drake/systems/analysis/test/stiff_double_mass_spring_system.h, line 90 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KlddxGMBVN4NKC9eVk4:-KlddxGMBVN4NKC9eVk5:b9ocjzz) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/stiff_double_mass_spring_system.h#L90)):*
> ```objc
>   /// Sets the initial conditions for the system.
>   /// The first mass will be located at x1 = 0.5 and second will be located at
>   /// x2 = 1.5. No initial velocity is present.
> ```

BTW, worth mentioning, probably in the class' documentation, that the resting position is for `x(0) = 0` and`x(1) = 1`

---

*[drake/systems/analysis/test/stiff_double_mass_spring_system.h, line 106 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kldf0Mjc0q3NjOltMyT:-Kldf0Mjc0q3NjOltMyU:b-8zwc43) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/stiff_double_mass_spring_system.h#L106)):*
> ```objc
>   /// Gets the solution for the system with initial state defined at @p context,
>   /// returning the solution at time @p t, in @p state.
>   void GetSolution(const Context<T>& context, const T& t,
> ```

This seems to provide the analytical solution for a very special case, namely the mode in which the two masses are oscillating together as a single body of mass (m1 + m2).
Could you elaborate on when this condition happens?:
- Initial conditions?
- Zero damping.
- Zero initial stretching between m1 and m2?

Those are just my guesses.

---

*[drake/systems/analysis/test/stiff_double_mass_spring_system.h, line 118 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldeR7wcV_Zv6CA8tTi:-KldeR7wcV_Zv6CA8tTj:b-8k4kwm) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/stiff_double_mass_spring_system.h#L118)):*
> ```objc
>     // non-stiff spring).
>     const double omega = std::sqrt(get_spring_constants()(0) /
>         (get_mass()(0) + get_mass()(1)));
> ```

BTW, maybe worth mentioning this corresponds to the mode in which the two masses move together as a single body (without any stretching happening between them).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-:-KldhsI_AjJlFKLkx3ak:b-yaoyt2)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Review status: all files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[drake/examples/rod2d/BUILD, line 45 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kldhmd-wqBWc39Wn_Dz:-KleZF4pkoWcbnPJcV9B:b-vtq8n7) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/examples/rod2d/BUILD#L45)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

why this new dependency? 
</blockquote></details>

Removed (I can explain f2f if you care how it appeared!)

---

*[drake/systems/analysis/BUILD, line 107 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldZgJyDzn8BdazZ-0v:-KleZxqVweR00RB7NvKJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/BUILD#L107)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, line break not needed.
</blockquote></details>

Done.

---

*[drake/systems/analysis/BUILD, line 114 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldZiLJwz8pfGmlcRdK:-KleZyhVyu5xO2nXbONX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/BUILD#L114)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, line break not needed.
</blockquote></details>

Done.

---

*[drake/systems/analysis/implicit_euler_integrator-inl.h, line 381 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldWKpqaei5ZTxmuUhr:-Kleh-h87VCYboDTbvy5:b8nvre1) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/implicit_euler_integrator-inl.h#L381)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, this seems reasonable but I don't quite the comment you make here. Is this because theta below gets closer to one? (or even greater than one) due to fact that this norm becomes so small?
I wonder if this is what we want in the general case. If the units of `dx` correspond to large numerical values, then you might never achieve such a small norm even though for the units of the problem the norm of `dx` could be quite small. Would we need an absolute tolerance or something of the like here?
</blockquote></details>

BTW PTAL now.

---

*[drake/systems/analysis/implicit_euler_integrator-inl.h, line 405 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldZ4rapfH3ijtYuKXn:-Kle_V-E2_mZb201xLvc:b-ctwyao) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/implicit_euler_integrator-inl.h#L405)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, I just realize that here you return successfully without updating the context's continuous state with the newer (hopefully better) `xtplus` (as you do later on in line 414. I think you'd like to have that update here also before you return.
</blockquote></details>

Done. (Nice observation!)

---

*[drake/systems/analysis/implicit_euler_integrator-inl.h, line 473 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldYzkULqCC70h4ic6K:-Klea6VpoHONGeZsx482:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/implicit_euler_integrator-inl.h#L473)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, odd indentation.
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 34 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldbtJDIYTmCktC8cq1:-KleaJsoUdGeI_jztysn:b-frmkyo) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L34)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Is the solution for t→∞ independent of the initial conditions?
I see the value of showing the integrator can go this far, but this test doesn't seem to test the accuracy of the solver (since any ok solver that is stable would be able to find the steady state solution).
</blockquote></details>

It's not completely independent, but I'm unsure of the degree of the dependence. You're right that this doesn't stress the accuracy of the solver, but then again, this is a first-order integrator! In other words, I'm more concerned that the integrator can find an approximate solution without taking forever (which is not the case for an explicit integrator). 

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 37 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kldb5AJ9firzQL_Z1tQ:-KlebAb5ifzZxHbf7lZg:b-8rs0p8) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L37)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, notice that for this tolerance, the reference solution could've been taken to be `[0, 0 ,1]` and the test would still pass. This might indicate that a more stressful test would be one where the solution is non-trivial, like for instance at `t = 1`?
</blockquote></details>

BTW I think getting the solution to an accuracy of 5e-5 on this problem is pretty damn great! Absolute error tolerances for the integrators in the table provided in the link (all of which were much higher order) were not *that* much better.

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 80 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldfXMgdv6YmvPC-blB:-KlecEzaHIjiT2NMXrD9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L80)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, `double_context_` could for a someone like me be confused with the scalar type. Maybe rename to `double_spring_context_`?
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 254 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kldfo63Nvks5Uit6ZoF:-KlecNMpCuqtMVnIuhW6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L254)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, **an** undamped
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 254 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldfyNFYLEtvznpxuKI:-KlecU9b7NIwoKqOM6Pu:b-qu6n94) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L254)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

> The solution of this system should be identical ....

Are you thinking of a particular mode being excited here? or are did you mean to say that the solution **approximates** (not identical) that of the two masses moving together?

</blockquote></details>

Good point. Done.

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 261 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldgEFmXGcRFY-BbEu2:-Kled69ynABEE6kdNbIp:bojkbvt) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L261)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

could we use a smaller number? is this large because the target accuracy is large? or because the approximation to the two masses moving together is only good to this precision?
</blockquote></details>

It's only good to that precision. Actually, I had to bump the accuracy up even further (just now) to guarantee even that.

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 298 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kldhb-mKwRRGpD3-81d:-Kled1XlzYMXSn121p3k:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L298)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

not used?
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/robertson_system.h, line 9 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kld_ZmXeHa-sz0F5m6i:-Kledk80qjMjF-YRxWvz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/robertson_system.h#L9)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, all throughout Drake we would've done:
```c++
namespace drake {
namespace systems {
namespace analysis {
namespace test {
```

I think worth trying to match wath everyone else does in Drake for consistency. 

</blockquote></details>

Done.

---

*[drake/systems/analysis/test/robertson_system.h, line 13 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kld_3wpYW-i2GE_yLDo:-Klef5MiVGH3MlMaVX5O:b-gb7vct) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/robertson_system.h#L13)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, I like this reference a lot  but I wonder if the site would be stable enough to have it here as a link. If the link dies then we lose this reference. Would there be a journal paper we can cite? could you push the pdf? Maybe worth complimenting stating the system of equations for this system and stating why is complex to solve (stiff, etc)
Ah! maybe we can cite the reference mentioned in that pdf? (H.H. Robertson in 1966 [Rob66])

</blockquote></details>

BTW The Hairer reference will not go away, and it's not unusual for us to cite web links. Yes, it's a danger that they'll go away, but AFAIK, I have yet to see us have a repository of web links stored in drake for referencing.

---

*[drake/systems/analysis/test/stiff_double_mass_spring_system.h, line 14 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldcTIjeoa6RgbSM3Tk:-KlekjJd0ZdzTQInht1X:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/stiff_double_mass_spring_system.h#L14)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, a brief statement of the equations would be useful here, even in second order form if that makes them shorter.
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/stiff_double_mass_spring_system.h, line 37 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KlddXyAN155068HsJGi:-KlegDMRXQXHzhk1v5AH:bbn8sxc) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/stiff_double_mass_spring_system.h#L37)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, space around `*` operator.
</blockquote></details>

BTW Strange- it's magically fixed in my version.

---

*[drake/systems/analysis/test/stiff_double_mass_spring_system.h, line 90 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KlddxGMBVN4NKC9eVk4:-KleiNE4DMaD8QqsiQ7b:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/stiff_double_mass_spring_system.h#L90)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, worth mentioning, probably in the class' documentation, that the resting position is for `x(0) = 0` and`x(1) = 1`
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/stiff_double_mass_spring_system.h, line 106 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kldf0Mjc0q3NjOltMyT:-KlensUN5CV5C_z9U3zI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/stiff_double_mass_spring_system.h#L106)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This seems to provide the analytical solution for a very special case, namely the mode in which the two masses are oscillating together as a single body of mass (m1 + m2).
Could you elaborate on when this condition happens?:
- Initial conditions?
- Zero damping.
- Zero initial stretching between m1 and m2?

Those are just my guesses.

</blockquote></details>

Done.

---

*[drake/systems/analysis/test/stiff_double_mass_spring_system.h, line 118 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldeR7wcV_Zv6CA8tTi:-KlenxIYP-Aj6oh1jjMl:b2diod0) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/stiff_double_mass_spring_system.h#L118)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, maybe worth mentioning this corresponds to the mode in which the two masses move together as a single body (without any stretching happening between them).
</blockquote></details>

BTW Covered by the documentation I just added for this function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6228)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 34 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldbtJDIYTmCktC8cq1:-KljuSxCrUCn4h8I-WNw:b7pzjjk) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L34)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

It's not completely independent, but I'm unsure of the degree of the dependence. You're right that this doesn't stress the accuracy of the solver, but then again, this is a first-order integrator! In other words, I'm more concerned that the integrator can find an approximate solution without taking forever (which is not the case for an explicit integrator). 
</blockquote></details>

Sure, we could always add another test later if we need to.

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 80 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KldfXMgdv6YmvPC-blB:-Kljuk4DJY3rinXXGcCY:b-8oixji) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L80)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I dont think that now follows the style guide but I'll mark myself satisfied.

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 609 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kljv92xP5xmTf3E1zrB:-Kljv92yFqS-0cn3t0Fu:b-77zjt) ([raw file](https://github.com/robotlocomotion/drake/blob/25e37d6fcf1679ea512004bcd95583f8aba0b077/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L609)):*
> ```cpp
> 
>   // Integrate exactly one step.
>   integrator.IntegrateWithMultipleSteps(large_dt_);
> ```

BTW, why the change from fixed to variable step?

---

*[drake/systems/analysis/test/robertson_system.h, line 13 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kld_3wpYW-i2GE_yLDo:-KljvPkgJXAHrm6bBdQX:bgrcpro) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/robertson_system.h#L13)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW The Hairer reference will not go away, and it's not unusual for us to cite web links. Yes, it's a danger that they'll go away, but AFAIK, I have yet to see us have a repository of web links stored in drake for referencing.
</blockquote></details>

BTW, maybe worth to have, in addition, the H.H. Robertson in 1966 [Rob66] citation as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6228)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
+@sammy-tri for platform review per regular rotation please

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-:-KlsizIRvqPwHBdcVCKI:b-8ba2gu)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
+(status: curate commits before merging) 

---

Reviewed 5 of 6 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/systems/analysis/test/robertson_system.h, line 23 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KlszOhCsdKN-9dARHkg:-KlszOhDA0gdIhE2Bh9-:b-ptemqy) ([raw file](https://github.com/robotlocomotion/drake/blob/25e37d6fcf1679ea512004bcd95583f8aba0b077/drake/systems/analysis/test/robertson_system.h#L23)):*
> ```objc
>   }
> 
>   void DoCalcTimeDerivatives(const Context <T> &context,
> ```

https://github.com/RobotLocomotion/drake/blob/master/drake/doc/code_style_guide.rst#clarifications says that the '&' should be next to the type, not the variable name  (also more below)

---

*[drake/systems/analysis/test/stiff_double_mass_spring_system.h, line 45 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Klszj5jDRN9u99Od_wp:-Klszj5koXl-h6CRQH8k:bvt9kfw) ([raw file](https://github.com/robotlocomotion/drake/blob/25e37d6fcf1679ea512004bcd95583f8aba0b077/drake/systems/analysis/test/stiff_double_mass_spring_system.h#L45)):*
> ```objc
>   /// Evaluates the spring and damping forces, returning the forces on each
>   /// body.
>   Vector2 <T> EvalSpringDamperForces(const Context <T> &context) const {
> ```

https://github.com/RobotLocomotion/drake/blob/master/drake/doc/code_style_guide.rst#clarifications says that the '&' should be next to the type, not the variable name

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-:-Klt-rtLOHiepr8Epd8Z:b-s794vx)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/systems/analysis/test/robertson_system.h, line 13 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kld_3wpYW-i2GE_yLDo:-Klt3G5xzJxhzG20NaZ4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c00ab0aa5eb5d3ae69b7e4c440211ea1301774b1/drake/systems/analysis/test/robertson_system.h#L13)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, maybe worth to have, in addition, the H.H. Robertson in 1966 [Rob66] citation as well.
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/robertson_system.h, line 23 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KlszOhCsdKN-9dARHkg:-Klt3tNMep2DCGZxmOTl:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/25e37d6fcf1679ea512004bcd95583f8aba0b077/drake/systems/analysis/test/robertson_system.h#L23)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

https://github.com/RobotLocomotion/drake/blob/master/drake/doc/code_style_guide.rst#clarifications says that the '&' should be next to the type, not the variable name  (also more below)
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/stiff_double_mass_spring_system.h, line 45 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Klszj5jDRN9u99Od_wp:-Klt4Z3SjP_VvoPA2MSA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/25e37d6fcf1679ea512004bcd95583f8aba0b077/drake/systems/analysis/test/stiff_double_mass_spring_system.h#L45)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

https://github.com/RobotLocomotion/drake/blob/master/drake/doc/code_style_guide.rst#clarifications says that the '&' should be next to the type, not the variable name
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6228)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Review status: 1 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 609 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Kljv92xP5xmTf3E1zrB:-KltCQLEb8BDIvfYnXSP:bdtzfe1) ([raw file](https://github.com/robotlocomotion/drake/blob/25e37d6fcf1679ea512004bcd95583f8aba0b077/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L609)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, why the change from fixed to variable step?
</blockquote></details>

BTW This change was made to better perform the stated test intention (checking global accuracy).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6228)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 4 of 4 files at r9.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 14 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KltCVSDmaIDPTz3fi7t:-KltCVSEUwq-Yy-yrxRf:bet43h9) ([raw file](https://github.com/robotlocomotion/drake/blob/e55b2fa85948fc886cc0e29e5350ac19e175f02f/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L14)):*
> ```cpp
> namespace {
> 
> using drake::systems::ImplicitEulerIntegrator;
> ```

This line appears as if it has no effect, since the current namespace scope is drake::systems.

---

*[drake/systems/analysis/test/robertson_system.h, line 23 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KlszOhCsdKN-9dARHkg:-KltCseIYV8LqkrMyerD:buwp8d0) ([raw file](https://github.com/robotlocomotion/drake/blob/25e37d6fcf1679ea512004bcd95583f8aba0b077/drake/systems/analysis/test/robertson_system.h#L23)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

https://google.github.io/styleguide/cppguide.html#Horizontal_Whitespace also says no whitespace before the "<" in templates and casts (sorry, missed it first time around)

---

*[drake/systems/analysis/test/stiff_double_mass_spring_system.h, line 45 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-Klszj5jDRN9u99Od_wp:-KltDBmqtuQAKX9_jPkn:buwp8d0) ([raw file](https://github.com/robotlocomotion/drake/blob/25e37d6fcf1679ea512004bcd95583f8aba0b077/drake/systems/analysis/test/stiff_double_mass_spring_system.h#L45)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

https://google.github.io/styleguide/cppguide.html#Horizontal_Whitespace also says no whitespace before the "<" in templates and casts (sorry, missed it first time around)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6228)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/analysis/test/implicit_euler_integrator_test.cc, line 14 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KltCVSDmaIDPTz3fi7t:-KltFmXY_r5t_wwTx3Oe:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e55b2fa85948fc886cc0e29e5350ac19e175f02f/drake/systems/analysis/test/implicit_euler_integrator_test.cc#L14)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

This line appears as if it has no effect, since the current namespace scope is drake::systems.
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/robertson_system.h, line 23 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-KlszOhCsdKN-9dARHkg:-KltGJ1TNp-xjj-r90gp:bjwyywa) ([raw file](https://github.com/robotlocomotion/drake/blob/25e37d6fcf1679ea512004bcd95583f8aba0b077/drake/systems/analysis/test/robertson_system.h#L23)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

https://google.github.io/styleguide/cppguide.html#Horizontal_Whitespace also says no whitespace before the "<" in templates and casts (sorry, missed it first time around)
</blockquote></details>

Done. (silly CLion defaults)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6228)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending curation

---

Reviewed 3 of 3 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6228#-:-KltIm0CpVH8_3d_TRSS:bj16f0t)*
<!-- Sent from Reviewable.io -->
